File: python/046.py
#!/usr/bin/env python3
"""
Project Euler Problem 46: Goldbach's other conjecture

It was proposed by Christian Goldbach that every odd composite number
can be written as the sum of a prime and twice a square.
Find the smallest odd composite that cannot be written as such.
"""

from math import isqrt


def is_prime(n: int) -> bool:
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    return True


def is_goldbach(n: int) -> bool:
    """Check if n = prime + 2 * square."""
    for prime in range(3, n, 2):
        if not is_prime(prime):
            continue
        remaining = n - prime
        if remaining % 2 != 0:
            continue
        square = remaining // 2
        if isqrt(square) ** 2 == square:
            return True
    return False


def main():
    n = 9
    while True:
        if n % 2 == 1 and not is_prime(n) and not is_goldbach(n):
            print(n)
            break
        n += 2


if __name__ == "__main__":
    main()

File: c/046.c
#include <stdio.h>
#include <stdbool.h>

bool is_prime(int n) {
    if (n < 2) return false;
    if (n == 2) return true;
    if (n % 2 == 0) return false;
    for (int i = 3; i * i <= n; i += 2) {
        if (n % i == 0) return false;
    }
    return true;
}

int main(void) {
    /* For each odd composite, check if it can be written as prime + 2*square */
    for (int i = 3; i < 10000; i += 2) {
        if (is_prime(i)) continue;

        bool found = false;
        for (int s = 1; 2 * s * s < i; s++) {
            int remainder = i - 2 * s * s;
            if (remainder > 0 && is_prime(remainder)) {
                found = true;
                break;
            }
        }

        if (!found) {
            printf("%d\n", i);
            return 0;
        }
    }

    return 0;
}
